File: library/src/main/java/io/fotoapparat/photo/Photo.java
package io.fotoapparat.photo;

import java.util.Arrays;

/**
 * Taken photo.
 */
public class Photo {

    /**
     * Encoded image. Use {@link android.graphics.BitmapFactory#decodeByteArray(byte[], int, int)}
     * to decode it.
     */
    public final byte[] encodedImage;

    /**
     * Clockwise rotation relatively to screen orientation at the moment when photo was taken. To
     * display the photo in a correct orientation it needs to be rotated counter clockwise by this
     * value.
     */
    public final int rotationDegrees;

    public Photo(byte[] encodedImage,
                 int rotationDegrees) {
        this.encodedImage = encodedImage;
        this.rotationDegrees = rotationDegrees;
    }

    /**
     * @return empty {@link Photo}.
     */
    public static Photo empty() {
        return new Photo(new byte[0], 0);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Photo photo = (Photo) o;

        return rotationDegrees == photo.rotationDegrees
                && Arrays.equals(encodedImage, photo.encodedImage);
    }

    @Override
    public int hashCode() {
        int result = Arrays.hashCode(encodedImage);
        result = 31 * result + rotationDegrees;
        return result;
    }

}

File: fotoapparat/src/main/java/io/fotoapparat/result/Photo.kt
package io.fotoapparat.result

import android.graphics.BitmapFactory
import java.util.Arrays

/**
 * Taken photo.
 */
data class Photo(

        /**
         * Encoded image. Use [android.graphics.BitmapFactory.decodeByteArray]
         * to decode it.
         */
        @JvmField
        val encodedImage: ByteArray,

        /**
         * Clockwise rotation relatively to screen orientation at the moment when photo was taken. To
         * display the photo in a correct orientation it needs to be rotated counter clockwise by this
         * value.
         */
        @JvmField
        val rotationDegrees: Int
) {

    /**
     * The height of the photo.
     */
    val height: Int
        get() = decodedBounds.outHeight

    /**
     * The width of the photo.
     */
    val width: Int
        get() = decodedBounds.outWidth

    private val decodedBounds by lazy {
        BitmapFactory.Options().apply {
            inJustDecodeBounds = true
            BitmapFactory.decodeByteArray(encodedImage, 0, encodedImage.size, this)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Photo

        if (!Arrays.equals(encodedImage, other.encodedImage)) return false
        if (rotationDegrees != other.rotationDegrees) return false

        return true
    }

    override fun hashCode(): Int {
        var result = Arrays.hashCode(encodedImage)
        result = 31 * result + rotationDegrees
        return result
    }

    override fun toString(): String =
        "Photo(encodedImage=ByteArray(${encodedImage.size}) rotationDegrees=$rotationDegrees)"

    companion object {
        /**
         * @return empty [Photo].
         */
        internal fun empty(): Photo = Photo(encodedImage = ByteArray(0), rotationDegrees = 0)
    }

}
